### PR TITLE
Add `paybackLoan` dftx + tx builder

### PR DESCRIPTION
**Pull request by canonbrother**
#### What kind of PR is this?:

/kind feature

#### Which issue(s) does this PR fixes?:

Fixes part of #532

### CODE REVIEW

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_loans/LoanPayback.test.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    const buffer = toBuffer(stack)
+    expect(buffer.toString('hex')).toStrictEqual(hex)
+    expect((stack[1] as OP_DEFI_TX).tx.type).toStrictEqual(0x48)
+  })
+})
+
+const header = '6a4c554466547848' // OP_RETURN(0x6a) OP_PUSHDATA1(0x4c) (length 85 = 0x55) CDfTx.SIGNATURE(0x44665478) CLoanPayback.OP_CODE(0x48)
+// LoanPayback.vaultId[LE](0x9199d45bdd1a746f4c899b06b79fadbf61cf88078004cdbc72ab8534406852f9)
+// LoanPayback.from(0x16001476060adc229ddaef9e2fea605e3c42954f8605b5)
+// LoanPayback.amount(0x0202000000006d7c4d00000000030000000046c32300000000)
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_loans/LoanPayback.test.ts:52]:
```suggestion
// LoanPayback.tokenAmounts(0x0202000000006d7c4d00000000030000000046c32300000000)
```

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+    vaultId: bobLiqVaultId,
+    amounts: '100@UBER'
+  })
+  await bob.generate(1)
+  await tGroup.waitForSync()
+
+  // liquidated: true
+  await alice.rpc.oracle.setOracleData(oracleId, timestamp, { prices: [{ tokenAmount: '100000@UBER', currency: 'USD' }] })
+  await alice.generate(1)
+  await tGroup.waitForSync()
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts:199]: I think you need `waitForSync` when you are calling `bob` next. 
You may remove `await tGroup.waitForSync()`

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+      tokenAmounts: [{ token: 2, amount: new BigNumber(30) }]
+    }, bobColScript)
+
+    const promise = sendTransaction(bob.container, txn)
+    await expect(promise).rejects.toThrow(DeFiDRpcError)
+    await expect(promise).rejects.toThrow(`Vault with id ${emptyVaultId} has no collaterals`)
+  })
+
+  it('should not loanPayback on liquidation vault', async () => {
+    await tGroup.get(0).generate(6)
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts:727]:
```suggestion
    await alice.generate(6)
```

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+      const interests = await bob.rpc.loan.getInterest('scheme')
+      const height = await bob.container.getBlockCount()
+      const tslaInterestPerBlock = (netInterest * 40) / (365 * blocksPerDay) //  netInterest * loanAmt / 365 * blocksPerDay
+      const tslaInterestTotal = tslaInterestPerBlock * (height - tslaLoanHeight)
+      expect(interests[0].interestPerBlock.toFixed(8)).toStrictEqual(tslaInterestPerBlock.toFixed(8))
+      expect(interests[0].totalInterest.toFixed(8)).toStrictEqual(tslaInterestTotal.toFixed(8))
+    }
+
+    await fundEllipticPair(bob.container, bProviders.ellipticPair, 10)
+    const bobColScript = P2WPKH.fromAddress(RegTest, bobColAddr, P2WPKH).getScript()
```
Comment (jingyi2811) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts:371]:
```
await fundEllipticPair(bob.container, bProviders.ellipticPair, 10)
const bobColScript = P2WPKH.fromAddress(RegTest, bobColAddr, P2WPKH).getScript()
```

Is that possible to move this 2 lines to beforeEach? So you do not need to duplicate the code in every test item

**Review by surangap — COMMENTED** [review]
loanPayback => paybackloan
Refactor accordingly 😅

**Comment by canonbrother** [comment]
> loanPayback => paybackloan Refactor accordingly sweat_smile

sure.. will do... because the ain side still remain loanPayback on raw side.. that's the reason i didn't rename it

**Comment by surangap** [comment]
> > loanPayback => paybackloan Refactor accordingly sweat_smile
> 
> sure.. will do... because the ain side still remain loanPayback on raw side.. that's the reason i didn't rename it

just saw that ain side still use `loanPayback` as message name.
Now i'm not sure. @fuxingloh any comment?

**Comment by fuxingloh** [comment]
> > > loanPayback => paybackloan Refactor accordingly sweat_smile
> > 
> > 
> > sure.. will do... because the ain side still remain loanPayback on raw side.. that's the reason i didn't rename it
> 
> just saw that ain side still use `loanPayback` as message name. Now i'm not sure. @fuxingloh any comment?

Yeah, I think we can and should strive to keep the consistency on our side.

**Comment by fuxingloh** [comment]
Similar stance on whale engineering, if it has nothing to do with consensus and purely for naming purposes, we do our best to keep to our own consistency and paradigm.

File: packages/jellyfish-transaction/src/script/dftx/dftx_loans.ts [hunk 5]
```diff
@@ -262,3 +262,29 @@ export class CTakeLoan extends ComposableBuffer<TakeLoan> {
...
+ */
+export class CPaybackLoan extends ComposableBuffer<PaybackLoan> {
+  static OP_CODE = 0x48 // 'H'
+  static OP_NAME = 'OP_DEFI_TX_PAYBACK_LOAN'
+
+  composers (lp: PaybackLoan): BufferComposer[] {
+    return [
+      ComposableBuffer.hexBEBufferLE(32, () => lp.vaultId, v => lp.vaultId = v),
+      ComposableBuffer.single<Script>(() => lp.from, v => lp.from = v, v => new CScript(v)),
+      ComposableBuffer.varUIntArray(() => lp.tokenAmounts, v => lp.tokenAmounts = v, v => new CTokenBalance(v))
```
Comment (surangap) [inline packages/jellyfish-transaction/src/script/dftx/dftx_loans.ts:287]:
```suggestion
  composers (pl: PaybackLoan): BufferComposer[] {
    return [
      ComposableBuffer.hexBEBufferLE(32, () => pl.vaultId, v => pl.vaultId = v),
      ComposableBuffer.single<Script>(() => pl.from, v => pl.from = v, v => new CScript(v)),
      ComposableBuffer.varUIntArray(() => pl.tokenAmounts, v => pl.tokenAmounts = v, v => new CTokenBalance(v))
```

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+  await alice.generate(1)
+  await tGroup.waitForSync()
+
+  await bob.rpc.loan.takeLoan({
+    vaultId: bobVaultId,
+    to: bobColAddr,
+    amounts: '40@TSLA'
+  })
+  tslaLoanHeight = await bob.container.getBlockCount()
+  await bob.generate(1)
```
Comment (surangap) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts:261]:
```suggestion
  await bob.generate(1)
  tslaLoanHeight = await bob.container.getBlockCount()
```

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts [hunk 7]
```diff
@@ -0,0 +1,770 @@
...
+  it('should paybackLoan', async () => {
+    await alice.rpc.account.sendTokensToAddress({}, { [bobColAddr]: ['5@TSLA'] })
+    await alice.generate(1)
+    await tGroup.waitForSync()
+
+    {
+      const interests = await bob.rpc.loan.getInterest('scheme')
+      const height = await bob.container.getBlockCount()
+      const tslaInterestPerBlock = (netInterest * 40) / (365 * blocksPerDay) //  netInterest * loanAmt / 365 * blocksPerDay
+      const tslaInterestTotal = tslaInterestPerBlock * (height - tslaLoanHeight)
```
Comment (surangap) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_loan_payback.test.ts:294]:
```suggestion
      const tslaInterestTotal = tslaInterestPerBlock * (height + 1 - tslaLoanHeight)
```

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_payback_loan.test.ts [hunk 8]
```diff
@@ -0,0 +1,770 @@
...
+      from: bobColScript,
+      tokenAmounts: [{ token: 2, amount: new BigNumber(30) }]
+    }, bobColScript)
+
+    const promise = sendTransaction(bob.container, txn)
+    await expect(promise).rejects.toThrow(DeFiDRpcError)
+    await expect(promise).rejects.toThrow('Cannot payback loan on vault under liquidation')
+  })
+
+  it('should not paybackLoan with arbitrary utxo', async () => {
```
Comment (surangap) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_payback_loan.test.ts:712]: utxo?
```suggestion
  it('should not paybackLoan with invalid auth', async () => {
```
?

Comment (canonbrother) [reply packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_payback_loan.test.ts]: ohh.. that's means using the other utxo which is not owned by user
yea `invalid auth` is more straight to the point

**Review by surangap — APPROVED** [review]

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts [hunk 9]
```diff
@@ -157,4 +158,19 @@ export class TxnBuilderLoans extends P2WPKHTxnBuilder {
...
   }
+
+  /**
+   * PaybackLoan to vault transaction.
+   *
+   * @param {PaybackLoan} paybackLoan txn to create
+   * @param {Script} changeScript to send unspent to after deducting the (converted + fees)
+   * @returns {Promise<TransactionSegWit>}
+   */
+
```
Comment (fuxingloh) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:169]:
```suggestion
```

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**